### DISCUSSION TITLE
How do I set "Use Vcpkg" to "NO" in Visual Studio property page widely or in a CMake manner?

### DISCUSSION THREAD

**Discussion by zchrissirhcz** [microsoft/vcpkg#19149] [Q&A]
Hi, vcpkg Team

I write C++ in Visual Studio, by first write CMakeLists.txt then generate .sln file, without using vcpkg.

Yesterday I  just installed vcpkg with Visual Studio integration via:
```
vcpkg integrate install
```

After the integration installation, there is `vcpkg` item in existing/new created Visual Project's property page, and `Use Vcpkg` is "YES" on default widely:

![image](https://user-images.githubusercontent.com/3831847/107469146-2dac3080-6ba4-11eb-8ab8-e7811e4b8233.png)

I know vcpkg is powerful and convenient from the readme page, however it may mess up my dependencies 

For example,
>compiling OpenCV imgcodecs module with zlib static library from source, however it automatically linking to vcpkg installed zlib dll file, which is not as expected.
(https://github.com/opencv/opencv/issues/17051)

My question: I would like to still keep vcpkg Visual Studio integration item/page in project property, and **is it possible to set `Use Vcpkg` to "NO" widely? And is it possible to just set it to "NO" for one specific project in a CMake manner? i.e. passing one option to disable / not use `vcpkg` temporarily? If possible then how to config that?**

Thanks.

**Comment by zchrissirhcz** [comment]
Ok just let me hack it...

Open `vcpkg.props` file (e.g. `E:\soft\vcpkg\scripts\buildsystems\msbuild\vcpkg.props`), 
modify the line:
```xml
<VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
```

to
```xml
<VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">false</VcpkgEnabled>
```

Then re-open Visual Studio, will see `Use Vcpkg` default to "NO" widely:

![image](https://user-images.githubusercontent.com/3831847/107472730-72d36100-6baa-11eb-9d91-fc8a4a7b714f.png)

**Comment by JackBoosY** [comment]
We recommend using all internal ports of vcpkg. For external libraries, you can use target_link_libraries to specify the location of use.

**Comment by zchrissirhcz** [comment]
@JackBoosY Good suggestion, and I think I can obey this suggestion for my personal project.

After a while of search, I figured out I can use this:
```cmake
set_target_properties(testbed PROPERTIES VS_GLOBAL_VcpkgEnabled false)
```
which disable `Use Vcpkg` for currect target temporarily.

**Comment by JackBoosY** [comment]
I prefer my colleague @ras0219 @ras0219-msft  to answer you more.

**Comment by zchrissirhcz** [comment]
Previously, the expected value for `VS_GLOBAL_VcpkgEnabled` was `false`, as discussed in https://github.com/microsoft/vcpkg/pull/5064 .

However the latest version of vcpkg set it to `true`. This makes me confused.

**Comment by zchrissirhcz** [comment]
Seems that due to commit 14514508d , `vcpkg\scripts\buildsystems\msbuild\vcpkg.props` in created and since then, there is:
```
<VcpkgEnabled Condition="'$(VcpkgEnabled)' == ''">true</VcpkgEnabled>
```

**Comment by zchrissirhcz** [comment]
OK, now I know the conflict:
- in `vcpkg\scripts\buildsystems\msbuild\vcpkg.props`, `VcpkgEnabled` is set to `true`
- in `vcpkg\scripts\buildsystems\vcpkg.cmake`, `VS_GLOBAL_VcpkgEnabled` is set to `false`

I suggest change `vcpkg.props`'s true to false.

**Comment by zchrissirhcz** [comment]
@JackBoosY   @ras0219 @ras0219-msft
I've just created an PR to fix that conflict.

**Comment by ras0219-msft** [comment]
Hi @zchrissirhcz,

The intended behavior is that:
1. If you create an MSBuild project, VcpkgEnabled will be set to true automatically. You can set it to false via project properties.
2. If you create a CMakeLists.txt project using our toolchain and generate MSBuild projects, VcpkgEnabled will be set to false (you are expected to use `find_package()` and friends to get dependencies)
3. If you create a CMakeLists.txt project without using our toolchain and generate MSBuild projects, VcpkgEnabled will be incidentally set to true. You can disable this by setting `VS_GLOBAL_VcpkgEnabled` to false.

Please let us know which of these situations you are in and what problems you are facing with configuring vcpkg on or off

**Comment by zchrissirhcz** [comment]
Hi @ras0219-msft, 

Thank you for your detailed explanation. I was in situation 3, and didn't know these different situations in the very beginning.

The problem I was facing: I build the open source projejct [OpenCV](https://github.com/opencv/opencv), which will build zlib static lib from source, and build opencv_imgcodecs shared lib that (expected to) relies on this static zlib. However zlib1.dll / zlibd1.dll (from vcpkg directory) is copied and relied by the generated opencv_imgcodecs shared lib, which is unexpected. (Actually, zlib static lib is generated.) The installation of OpenCV package won't include zlib(d)1.dll, using this package will cause "zlib(d)1.dll not found" error on both my machine and other machine.

I think there are some gaps:

1. I didn't know what exactly does `vcpkg integrate install` do, since I just copy-and-paste-and-run that command, without immediately creating either a MSBuild project nor a "CMakeLists.txt + no-vcpkg-toolchain + MSBuild" project. I even didn't notice there will be a `vcpkg` config item page in Visual Studio project property page. **If put that config screenshot in the readme.md page, it would better**.

2. `VS_GLOBAL_VcpkgEnabled` is target level config, `vcpkg integrate install/remove` is all VS solution level, can I turn on/off vcpkg integration for each VS solution?  For example, when building a multi-target projects, like OpenCV, I have to **turn off `VcpkgEnabled` for each target, which is inconvenient, turn off `VcpkgEnabled` for the whole solution (all the targets under this VS solution) is expected**, while keeps other VS solution can still use vcpkg.

**Comment by zchrissirhcz** [comment]
When I create PR #16167 , I thought situation 2 and situation 3 should keep same expectation, VcpkgEnabled set to false.

Now I change my mind. I think this would be better:
- situation 2, `VcpkgEnabled` set to true on default, since passing vcpkg's toolchain means using vcpkg's packages
- situation 3, `VcpkgEnabled` set to false on default, since not passing vcpkg's toolchain, should behave the same as "vcpkg integrate remove".

Pushed a new commit to #16167 for my current expectation.

**Comment by zchrissirhcz** [comment]
For situation 3, let me put an example, to make it clear why "VcpkgEnabled default to false is expected".

It is a hello-world program that dependends on zlib, but in its CMakeLists.txt, doesn't specify that dependency. When building this program, doesn't specify vcpkg toolchain file.  Thus, **it is expected that there will be an link error**. However, vcpkg integration brokes this. It can build and run.

Reproduceable stpes / code are in the following.

## Commands:
```
vcpkg integrate install
vcpkg install zlib:x64-windows
```

## Code files
```
$ tree
.
|-- CMakeLists.txt
|-- build
|   `-- vs2019-x64.cmd
`-- testbed.cpp
```

testbed.cpp:
```c++
#include <iostream>
#include <string>
#include <zlib.h>
#include <zconf.h>

using std::cout;
using std::endl;

int main(int argc, char* argv[])
{
    unsigned char szSrc[] = "test the compression and uncompression of zlib.";
    unsigned long nSrcLen = sizeof(szSrc);
    unsigned char szZip[1024] = { 0 };
    unsigned long nZipLen = 1024;
    compress(szZip, &nZipLen, szSrc, nSrcLen);
    unsigned char szUnZip[1024] = { 0 };
    unsigned long nUnZipLen = 1024;
    uncompress(szUnZip, &nUnZipLen, szZip, nZipLen);
    cout << "Src:" << szSrc << ", len:" << nSrcLen << endl;
    cout << "Zip:" << szZip << ", len:" << nZipLen << endl;
    cout << "UnZip:" << szUnZip << ", len:" << nUnZipLen << endl;

    return 0;
}
```

CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 3.15)

project(zlib_test)

add_executable(testbed
    testbed.cpp
)
```

build/vs2019-x64.cmd:
```
@echo off

set BUILD_DIR=vs2019-x64
if not exist %BUILD_DIR% md %BUILD_DIR%
cd %BUILD_DIR%
cmake ../.. -G "Visual Studio 16 2019" -A x64

cd ..
pause
```

## What's more

If I just specify my target `testbed` linking to zlib in CMakeLists.txt, which zlib (my specified, and vcpkg installed one) will be linked? This will be an linking order dependent issue.

In a more badly case, the previously mentioned OpenCV imgcodecs shared library example, the actual link order is:
(note OpenCV will build both zlib and libpng static libraries)
```
imgcodecs --> zlib --> libpng       /// link case 1
```

But the correct link order should be:
```
imgcodecs --> libpng --> zlib     /// link case 2
```

And as mentioned in https://docs.microsoft.com/en-us/cpp/build/reference/link-input-files?view=msvc-160 ,
>Object files on the command line are processed in the order they appear on the command line. Libraries are searched in command line order as well, with the following caveat: Symbols that are unresolved when bringing in an object file from a library are searched for in that library first, and then the following libraries from the command line and /DEFAULTLIB (Specify Default Library) directives, and then to any libraries at the beginning of the command line.

Thus, the link case 1 can still generate imgcodecs shared library, but it may linking to vcpkg installed zlib, not the build-from-source static zlib. (This was my guess).

To validate this guess, put `/VERBOSE` to link command line option, and gains:
```
...

5>    正在搜索 D:\soft\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\zlibd.lib:
5>      已找到 inflateEnd
5>        已在 libpngd.lib(pngread.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 deflateEnd
5>        已在 libpngd.lib(pngwrite.obj) 中引用
5>        已在 libpngd.lib(pngwutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 inflateReset
5>        已在 libpngd.lib(png.obj) 中引用
5>        已在 libpngd.lib(pngrutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 adler32
5>        已在 libpngd.lib(png.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 crc32
5>        已在 libpngd.lib(png.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 inflate
5>        已在 libpngd.lib(pngrutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 inflateReset2
5>        已在 libpngd.lib(pngrutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 inflateInit2_
5>        已在 libpngd.lib(pngrutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 inflateValidate
5>        已在 libpngd.lib(pngrutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 deflate
5>        已在 libpngd.lib(pngwutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 deflateReset
5>        已在 libpngd.lib(pngwutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 deflateInit2_
5>        已在 libpngd.lib(pngwutil.obj) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 __IMPORT_DESCRIPTOR_zlibd1
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)
5>      已找到 zlibd1_NULL_THUNK_DATA
5>        已在 zlibd.lib(zlibd1.dll) 中引用
5>        已加载 zlibd.lib(zlibd1.dll)

...

```

Which means the symbol `inflateEnd`, called in  libpngd.lib(pngread.obj),  is find in library file `D:\soft\vcpkg\scripts\buildsystems\msbuild\..\..\..\installed\x64-windows\debug\lib\zlibd.lib:`. Well, this makes the conflict.

**Comment by zchrissirhcz** [comment]
By searching "what exactly link command is used for my conflict zlib OpenCV build", I find `D:\work\opencv-dbg\build\vs2019-x64\modules\imgcodecs\opencv_imgcodecs.dir\Debug\opencv_imgcodecs.tlog\link.command.1.tlog`, whose content:

```
^D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\BITSTRM.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\EXIF.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\GRFMT_BASE.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\GRFMT_BMP.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\GRFMT_PNG.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\LOADSAVE.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\OPENCV_IMGCODECS_MAIN.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\RGBE.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\UTILS.OBJ|D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\MODULES\IMGCODECS\OPENCV_IMGCODECS.DIR\DEBUG\VS_VERSION.RES
/OUT:"D:\WORK\OPENCV-DBG\BUILD\VS2019-X64\BIN\DEBUG\OPENCV_IMGCODECS451D.DLL" /INCREMENTAL /NOLOGO /LIBPATH:"D:\SOFT\VCPKG\SCRIPTS\BUILDSYSTEMS\MSBUILD\..\..\..\INSTALLED\X64-WINDOWS\DEBUG\LIB" /LIBPATH:"D:\SOFT\VCPKG\SCRIPTS\BUILDSYSTEMS\MSBUILD\..\..\..\INSTALLED\X64-WINDOWS\DEBUG\LIB\MANUAL-LINK" ..\..\LIB\DEBUG\OPENCV_CORE451D.LIB ..\..\3RDPARTY\LIB\DEBUG\ZLIBD.LIB ..\..\3RDPARTY\LIB\DEBUG\LIBPNGD.LIB ..\..\3RDPARTY\LIB\DEBUG\ZLIBD.LIB KERNEL32.LIB USER32.LIB GDI32.LIB WINSPOOL.LIB SHELL32.LIB OLE32.LIB OLEAUT32.LIB UUID.LIB COMDLG32.LIB ADVAPI32.LIB "D:\SOFT\VCPKG\SCRIPTS\BUILDSYSTEMS\MSBUILD\..\..\..\INSTALLED\X64-WINDOWS\DEBUG\LIB\*.LIB" /NODEFAULTLIB:LIBC /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"D:/WORK/OPENCV-DBG/BUILD/VS2019-X64/BIN/DEBUG/OPENCV_IMGCODECS451D.PDB" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"D:/WORK/OPENCV-DBG/BUILD/VS2019-X64/LIB/DEBUG/OPENCV_IMGCODECS451D.LIB" /MACHINE:X64  /machine:x64 /VERBOSE /DLL OPENCV_IMGCODECS.DIR\DEBUG\VS_VERSION.RES
OPENCV_IMGCODECS.DIR\DEBUG\LOADSAVE.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\UTILS.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\GRFMT_BASE.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\GRFMT_BMP.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\GRFMT_PNG.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\BITSTRM.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\RGBE.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\EXIF.OBJ
OPENCV_IMGCODECS.DIR\DEBUG\OPENCV_IMGCODECS_MAIN.OBJ
```

Among them, the key content is **LIBPATH**. 

**'/LIBPATH:"D:\\SOFT\\VCPKG\\SCRIPTS\\BUILDSYSTEMS\\MSBUILD\\..\\..\\..\\INSTALLED\\X64-WINDOWS\\DEBUG\\LIB" '**

Let me format it as:
```
('/OUT:"D:\\WORK\\OPENCV-DBG\\BUILD\\VS2019-X64\\BIN\\DEBUG\\OPENCV_IMGCODECS451D.DLL" '
 '/INCREMENTAL /NOLOGO '
 '/LIBPATH:"D:\\SOFT\\VCPKG\\SCRIPTS\\BUILDSYSTEMS\\MSBUILD\\..\\..\\..\\INSTALLED\\X64-WINDOWS\\DEBUG\\LIB" '
 '/LIBPATH:"D:\\SOFT\\VCPKG\\SCRIPTS\\BUILDSYSTEMS\\MSBUILD\\..\\..\\..\\INSTALLED\\X64-WINDOWS\\DEBUG\\LIB\\MANUAL-LINK" '
 '..\\..\\LIB\\DEBUG\\OPENCV_CORE451D.LIB '
 '..\\..\x03RDPARTY\\LIB\\DEBUG\\ZLIBD.LIB '
 '..\\..\x03RDPARTY\\LIB\\DEBUG\\LIBPNGD.LIB '
 '..\\..\x03RDPARTY\\LIB\\DEBUG\\ZLIBD.LIB KERNEL32.LIB USER32.LIB GDI32.LIB '
 'WINSPOOL.LIB SHELL32.LIB OLE32.LIB OLEAUT32.LIB UUID.LIB COMDLG32.LIB '
 'ADVAPI32.LIB '
 '"D:\\SOFT\\VCPKG\\SCRIPTS\\BUILDSYSTEMS\\MSBUILD\\..\\..\\..\\INSTALLED\\X64-WINDOWS\\DEBUG\\LIB\\*.LIB" '
 '/NODEFAULTLIB:LIBC /MANIFEST /MANIFESTUAC:"level=\'asInvoker\' '
 'uiAccess=\'false\'" /manifest:embed /DEBUG '
 '/PDB:"D:/WORK/OPENCV-DBG/BUILD/VS2019-X64/BIN/DEBUG/OPENCV_IMGCODECS451D.PDB" '
 '/SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT '
 '/IMPLIB:"D:/WORK/OPENCV-DBG/BUILD/VS2019-X64/LIB/DEBUG/OPENCV_IMGCODECS451D.LIB" '
 '/MACHINE:X64  /machine:x64 /VERBOSE /DLL '
 "OPENCV_IMGCODECS.DIR\\DEBUG\\VS_VERSION.RES'")
```

This LIBPATH specification makes the linker mislinked vcpkg's zlib dll.

**Comment by JackBoosY** [comment]
cc @ras0219 @ras0219-msft for response.

**Comment by JackBoosY** [comment]
I suggest you to use manifest mode to avoid this. Because in manifest mode, all installed pacakges are separated.
See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md).

**Reply by zchrissirhcz** [reply]
I don't think manifest mode is good enough to solve the proposed issue. Many people never heard of manifest mode, they have to read the docs. But before the know the `manifest` word, how can they easily get that document?

My proposed solution is to change the not-good-default-behaviour.

**Reply by zchrissirhcz** [reply]
I'm currently busy for other stuffs, didn't test if the mentioned manifest mode is working.

**Reply by JackBoosY** [reply]
In the future, most of the users will use manifest mode instead of classic mode.
See Chinese docs:
https://www.cnblogs.com/vcpkg/p/15019867.html
https://www.cnblogs.com/vcpkg/p/15019899.html

**Reply by Timie** [reply]
In the meantime, can we at least get the VCPKG to do a reasonable default behaviour, as @zchrissirhcz proposed?

Right now, I need to edit most of CMake projects across the company even though they do not use VCPKG, because of the incorrect default VCPKG integration behaviour. It is very strange that we are required to perform VCPKG-specific target-level configuration in CMake only if you are NOT using vcpkg. One should not even care about VCPKG if you opted out from it.

**Reply by brad-t-moore** [reply]
I don't know if this is the same issue, but I just had the darndest time debugging a hidden linking command in Visual Studio involving vcpkg.  It was adding something like "vcpkg/installed/lib/*.lib" which was wreaking havoc on my build because I had newer versions of libraries I was trying to link against.  It also wasn't shown in the Linker command-line display in Visual Studio- I had to run the build process in Diagnostic mode and it was just dumb luck in grep'ing the output to find the issue (I had no idea vcpkg would even be involved in this particular build).  That's just wild.  The reality is that developers are going to have local compiled versions of libraries, different versions of libraries, etc. on the same machine.  This implicit centralized repo idea is a bad default (all of Python dev is going to local virtual envs and that is a bright future).  If anything, make it easy for developers to set up VCPKG in their search path for FindPackage calls in CMake, but make it explicit (explicit > mysteriously sometimes convenient).

**Reply by JackBoosY** [reply]
cc @BillyONeal

**Comment by emerth** [comment]
Oh this is a wonderful discussion. I have been beating my head against a related problem for a week, and the root cause proved to be the vcpkg.cmake VS_GLOBAL_VcpkgEnabled was true but the vcpkg.props VcpkgEnabled was false. My problematic project was using SLN generated by cmake. I was really starting to question my sanity :-)
